File: ZevClip/FileTransfer/FileTransferStateMachine.swift
import Foundation

enum FileTransferState: String, Codable, Equatable {
    case offered
    case awaitingAcceptance
    case accepted
    case transferring
    case interrupted
    case verifying
    case completed
    case declined
    case cancelled
    case failed

    var isTerminal: Bool {
        switch self {
        case .completed, .declined, .cancelled, .failed:
            return true
        case .offered, .awaitingAcceptance, .accepted, .transferring, .interrupted, .verifying:
            return false
        }
    }
}

enum FileTransferCancellationReason: String, Codable, Equatable {
    case cancelledBySender
    case cancelledByReceiver
    case userCancelled
    case appShutdown
}

enum FileTransferFailureReason: String, Codable, Equatable {
    case authenticationFailed
    case certificateMismatch
    case invalidManifest
    case invalidRange
    case hashMismatch
    case missingChunks
    case destinationChanged
    case resumeExpired
    case internalError
}

enum FileTransferStateMachineError: Error, LocalizedError, Equatable {
    case invalidInitialState
    case invalidTransition(from: FileTransferState, to: FileTransferState)
    case invalidProgress
    case invalidActiveStreamCount
    case incompleteTransfer

    var errorDescription: String? {
        switch self {
        case .invalidInitialState:
            return "A transfer state machine cannot start in a terminal state."
        case .invalidTransition(let from, let to):
            return "Illegal file transfer state transition: \(from.rawValue) -> \(to.rawValue)."
        case .invalidProgress:
            return "Transfer progress is invalid."
        case .invalidActiveStreamCount:
            return "Active stream count is invalid."
        case .incompleteTransfer:
            return "Cannot verify a transfer before all bytes are verified."
        }
    }
}

struct FileTransferProgress: Codable, Equatable {
    let totalBytes: Int64
    let verifiedBytes: Int64
    let activeStreamCount: Int

    init(totalBytes: Int64, verifiedBytes: Int64 = 0, activeStreamCount: Int = 0) throws {
        guard totalBytes >= 0,
              verifiedBytes >= 0,
              verifiedBytes <= totalBytes,
              (0...ZevLinkTransferProtocol.maxStreamCount).contains(activeStreamCount) else {
            throw FileTransferStateMachineError.invalidProgress
        }
        self.totalBytes = totalBytes
        self.verifiedBytes = verifiedBytes
        self.activeStreamCount = activeStreamCount
    }

    var remainingBytes: Int64 {
        totalBytes - verifiedBytes
    }

    var fractionComplete: Double {
        totalBytes == 0 ? 1.0 : Double(verifiedBytes) / Double(totalBytes)
    }

    var isComplete: Bool {
        verifiedBytes == totalBytes
    }

    func addingVerifiedBytes(_ byteCount: Int64) throws -> FileTransferProgress {
        guard byteCount >= 0 else {
            throw FileTransferStateMachineError.invalidProgress
        }
        return try FileTransferProgress(
            totalBytes: totalBytes,
            verifiedBytes: min(verifiedBytes + byteCount, totalBytes),
            activeStreamCount: activeStreamCount
        )
    }

    func withActiveStreams(_ activeStreams: Int) throws -> FileTransferProgress {
        guard (0...ZevLinkTransferProtocol.maxStreamCount).contains(activeStreams) else {
            throw FileTransferStateMachineError.invalidActiveStreamCount
        }
        return try FileTransferProgress(
            totalBytes: totalBytes,
            verifiedBytes: verifiedBytes,
            activeStreamCount: activeStreams
        )
    }
}

struct FileTransferStateSnapshot: Codable, Equatable {
    let state: FileTransferState
    let progress: FileTransferProgress
    let selectedStreamCount: Int
    let cancellationReason: FileTransferCancellationReason?
    let failureReason: FileTransferFailureReason?
}

struct FileTransferStateMachine {
    private(set) var state: FileTransferState
    private(set) var progress: FileTransferProgress
    private(set) var cancellationReason: FileTransferCancellationReason?
    private(set) var failureReason: FileTransferFailureReason?
    let selectedStreamCount: Int

    init(
        initialState: FileTransferState = .offered,
        totalBytes: Int64,
        selectedStreamCount: Int
    ) throws {
        guard !initialState.isTerminal else {
            throw FileTransferStateMachineError.invalidInitialState
        }
        guard (ZevLinkTransferProtocol.minStreamCount...ZevLinkTransferProtocol.maxStreamCount).contains(selectedStreamCount) else {
            throw FileTransferValidationError.invalidStreamCount
        }

        state = initialState
        progress = try FileTransferProgress(totalBytes: totalBytes)
        self.selectedStreamCount = selectedStreamCount
    }

    var snapshot: FileTransferStateSnapshot {
        FileTransferStateSnapshot(
            state: state,
            progress: progress,
            selectedStreamCount: selectedStreamCount,
            cancellationReason: cancellationReason,
            failureReason: failureReason
        )
    }

    mutating func persistOffer(requiresUserDecision: Bool) throws {
        try transition(
            to: requiresUserDecision ? .awaitingAcceptance : .accepted
        )
    }

    mutating func applyDecision(_ decision: FileTransferDecision) throws {
        switch decision {
        case .accept:
            try transition(to: .accepted)
        case .decline:
            try transition(to: .declined)
        }
    }

    mutating func markChunkRequestStarted(activeStreamCount: Int) throws {
        guard (ZevLinkTransferProtocol.minStreamCount...selectedStreamCount).contains(activeStreamCount) else {
            throw FileTransferStateMachineError.invalidActiveStreamCount
        }

        if state == .accepted || state == .interrupted {
            try transition(to: .transferring)
        } else if state != .transferring {
            throw FileTransferStateMachineError.invalidTransition(from: state, to: .transferring)
        }

        progress = try progress.withActiveStreams(activeStreamCount)
    }

    mutating func markChunkVerified(byteCount: Int64) throws {
        guard state == .transferring else {
            throw FileTransferStateMachineError.invalidTransition(from: state, to: .transferring)
        }
        progress = try progress.addingVerifiedBytes(byteCount)
    }

    mutating func markNoActiveConnections() throws {
        progress = try progress.withActiveStreams(0)
        if state == .accepted || state == .transferring {
            try transition(to: .interrupted)
        }
    }

    mutating func beginVerification() throws {
        guard progress.isComplete else {
            throw FileTransferStateMachineError.incompleteTransfer
        }
        try transition(to: .verifying)
    }

    mutating func complete() throws {
        try transition(to: .completed)
    }

    mutating func cancel(reason: FileTransferCancellationReason) throws {
        cancellationReason = reason
        try transition(to: .cancelled)
    }

    mutating func fail(reason: FileTransferFailureReason) throws {
        failureReason = reason
        try transition(to: .failed)
    }

    mutating func transition(to nextState: FileTransferState) throws {
        guard Self.canTransition(from: state, to: nextState) else {
            throw FileTransferStateMachineError.invalidTransition(from: state, to: nextState)
        }
        state = nextState
        if nextState.isTerminal {
            progress = try progress.withActiveStreams(0)
        }
    }

    static func canTransition(from: FileTransferState, to: FileTransferState) -> Bool {
        if from == to {
            return true
        }
        if from.isTerminal {
            return false
        }
        if to == .cancelled {
            return from != .declined && from != .completed
        }
        if to == .failed {
            return true
        }

        switch from {
        case .offered:
            return to == .awaitingAcceptance || to == .accepted
        case .awaitingAcceptance:
            return to == .accepted || to == .declined
        case .accepted:
            return to == .transferring || to == .interrupted || to == .verifying
        case .transferring:
            return to == .interrupted || to == .verifying
        case .interrupted:
            return to == .transferring
        case .verifying:
            return to == .completed
        case .completed, .declined, .cancelled, .failed:
            return false
        }
    }
}

File: android/app/src/main/java/com/zevclip/sender/filetransfer/FileTransferStateMachine.kt
package com.zevclip.sender.filetransfer

enum class FileTransferState(val wireValue: String, val isTerminal: Boolean = false) {
    OFFERED("offered"),
    AWAITING_ACCEPTANCE("awaitingAcceptance"),
    ACCEPTED("accepted"),
    TRANSFERRING("transferring"),
    INTERRUPTED("interrupted"),
    VERIFYING("verifying"),
    COMPLETED("completed", isTerminal = true),
    DECLINED("declined", isTerminal = true),
    CANCELLED("cancelled", isTerminal = true),
    FAILED("failed", isTerminal = true);

    companion object {
        fun fromWireValue(value: String): FileTransferState? {
            return entries.firstOrNull { it.wireValue == value }
        }
    }
}

enum class FileTransferCancellationReason(val wireValue: String) {
    CANCELLED_BY_SENDER("cancelledBySender"),
    CANCELLED_BY_RECEIVER("cancelledByReceiver"),
    USER_CANCELLED("userCancelled"),
    APP_SHUTDOWN("appShutdown")
}

enum class FileTransferFailureReason(val wireValue: String) {
    AUTHENTICATION_FAILED("authenticationFailed"),
    CERTIFICATE_MISMATCH("certificateMismatch"),
    INVALID_MANIFEST("invalidManifest"),
    INVALID_RANGE("invalidRange"),
    HASH_MISMATCH("hashMismatch"),
    MISSING_CHUNKS("missingChunks"),
    DESTINATION_CHANGED("destinationChanged"),
    RESUME_EXPIRED("resumeExpired"),
    INTERNAL_ERROR("internalError")
}

data class FileTransferProgress(
    val totalBytes: Long,
    val verifiedBytes: Long = 0L,
    val activeStreamCount: Int = 0
) {
    init {
        require(totalBytes >= 0L) { "Total bytes must not be negative." }
        require(verifiedBytes in 0L..totalBytes) { "Verified bytes must be within total bytes." }
        require(activeStreamCount in 0..ZevLinkTransferProtocol.MAX_STREAM_COUNT) {
            "Active stream count must be between 0 and 8."
        }
    }

    val remainingBytes: Long
        get() = totalBytes - verifiedBytes

    val fractionComplete: Double
        get() = if (totalBytes == 0L) 1.0 else verifiedBytes.toDouble() / totalBytes.toDouble()

    val isComplete: Boolean
        get() = verifiedBytes == totalBytes

    fun addingVerifiedBytes(byteCount: Long): FileTransferProgress {
        require(byteCount >= 0L) { "Verified byte count must not be negative." }
        val nextVerifiedBytes = (verifiedBytes + byteCount).coerceAtMost(totalBytes)
        return copy(verifiedBytes = nextVerifiedBytes)
    }

    fun withActiveStreams(activeStreams: Int): FileTransferProgress {
        require(activeStreams in 0..ZevLinkTransferProtocol.MAX_STREAM_COUNT) {
            "Active stream count must be between 0 and 8."
        }
        return copy(activeStreamCount = activeStreams)
    }
}

data class FileTransferStateSnapshot(
    val state: FileTransferState,
    val progress: FileTransferProgress,
    val selectedStreamCount: Int,
    val cancellationReason: FileTransferCancellationReason? = null,
    val failureReason: FileTransferFailureReason? = null
)

class FileTransferStateMachine(
    initialState: FileTransferState = FileTransferState.OFFERED,
    totalBytes: Long,
    selectedStreamCount: Int
) {
    var state: FileTransferState = initialState
        private set

    var progress: FileTransferProgress = FileTransferProgress(totalBytes = totalBytes)
        private set

    var cancellationReason: FileTransferCancellationReason? = null
        private set

    var failureReason: FileTransferFailureReason? = null
        private set

    init {
        require(selectedStreamCount in ZevLinkTransferProtocol.MIN_STREAM_COUNT..ZevLinkTransferProtocol.MAX_STREAM_COUNT) {
            "Selected stream count must be between 1 and 8."
        }
        require(!initialState.isTerminal) { "A new state machine cannot start in a terminal state." }
    }

    val selectedStreamCount: Int = selectedStreamCount

    val snapshot: FileTransferStateSnapshot
        get() = FileTransferStateSnapshot(
            state = state,
            progress = progress,
            selectedStreamCount = selectedStreamCount,
            cancellationReason = cancellationReason,
            failureReason = failureReason
        )

    fun persistOffer(requiresUserDecision: Boolean) {
        transitionTo(
            if (requiresUserDecision) {
                FileTransferState.AWAITING_ACCEPTANCE
            } else {
                FileTransferState.ACCEPTED
            }
        )
    }

    fun applyDecision(decision: FileTransferDecision) {
        when (decision) {
            FileTransferDecision.ACCEPT -> transitionTo(FileTransferState.ACCEPTED)
            FileTransferDecision.DECLINE -> transitionTo(FileTransferState.DECLINED)
        }
    }

    fun markChunkRequestStarted(activeStreamCount: Int) {
        require(activeStreamCount in ZevLinkTransferProtocol.MIN_STREAM_COUNT..selectedStreamCount) {
            "Active stream count must be between 1 and the selected stream count."
        }
        if (state == FileTransferState.ACCEPTED || state == FileTransferState.INTERRUPTED) {
            transitionTo(FileTransferState.TRANSFERRING)
        } else if (state != FileTransferState.TRANSFERRING) {
            error("Cannot start chunk upload while transfer is ${state.wireValue}.")
        }
        progress = progress.withActiveStreams(activeStreamCount)
    }

    fun markChunkVerified(byteCount: Long) {
        if (state != FileTransferState.TRANSFERRING) {
            error("Cannot verify chunk bytes while transfer is ${state.wireValue}.")
        }
        progress = progress.addingVerifiedBytes(byteCount)
    }

    fun markNoActiveConnections() {
        progress = progress.withActiveStreams(0)
        if (state == FileTransferState.ACCEPTED || state == FileTransferState.TRANSFERRING) {
            transitionTo(FileTransferState.INTERRUPTED)
        }
    }

    fun beginVerification() {
        require(progress.isComplete) { "Cannot verify a transfer before all bytes are verified." }
        transitionTo(FileTransferState.VERIFYING)
    }

    fun complete() {
        transitionTo(FileTransferState.COMPLETED)
    }

    fun cancel(reason: FileTransferCancellationReason) {
        cancellationReason = reason
        transitionTo(FileTransferState.CANCELLED)
    }

    fun fail(reason: FileTransferFailureReason) {
        failureReason = reason
        transitionTo(FileTransferState.FAILED)
    }

    fun transitionTo(nextState: FileTransferState) {
        if (!canTransition(state, nextState)) {
            error("Illegal file transfer state transition: ${state.wireValue} -> ${nextState.wireValue}.")
        }
        state = nextState
        if (nextState.isTerminal) {
            progress = progress.withActiveStreams(0)
        }
    }

    companion object {
        fun canTransition(from: FileTransferState, to: FileTransferState): Boolean {
            if (from == to) {
                return true
            }
            if (from.isTerminal) {
                return false
            }
            if (to == FileTransferState.CANCELLED) {
                return from != FileTransferState.DECLINED && from != FileTransferState.COMPLETED
            }
            if (to == FileTransferState.FAILED) {
                return true
            }

            return when (from) {
                FileTransferState.OFFERED ->
                    to == FileTransferState.AWAITING_ACCEPTANCE ||
                        to == FileTransferState.ACCEPTED
                FileTransferState.AWAITING_ACCEPTANCE ->
                    to == FileTransferState.ACCEPTED ||
                        to == FileTransferState.DECLINED
                FileTransferState.ACCEPTED ->
                    to == FileTransferState.TRANSFERRING ||
                        to == FileTransferState.INTERRUPTED ||
                        to == FileTransferState.VERIFYING
                FileTransferState.TRANSFERRING ->
                    to == FileTransferState.INTERRUPTED ||
                        to == FileTransferState.VERIFYING
                FileTransferState.INTERRUPTED ->
                    to == FileTransferState.TRANSFERRING
                FileTransferState.VERIFYING ->
                    to == FileTransferState.COMPLETED
                FileTransferState.COMPLETED,
                FileTransferState.DECLINED,
                FileTransferState.CANCELLED,
                FileTransferState.FAILED -> false
            }
        }
    }
}

